src/library/scala/collection/immutable/ChampHashMap.scala

```diff
@@ -205,7 +229,8 @@ private final class BitmapIndexedMapNode[K, +V](val dataMap: Int, val nodeMap: I
 
     if ((dataMap & bitpos) != 0) {
       val index = indexFrom(dataMap, mask, bitpos)
-      return (key, value) == this.getPayload(index)
+      // TODO test drive the change from == to eq here.
+      key == this.getKey(index) && (value.asInstanceOf[AnyRef] eq this.getValue(index).asInstanceOf[AnyRef])
```

Comment:
The only time we should `Object.equals` (ala `AnyRef.==`) when from `Map.hashCode,equals`.

 - to be conservative about not knowing whether it is a cheap or expensive operation, and
 - because the object identity is an observable, we the property that `map.update(k, v).apply(k) eq v` should hold.

Reply:
At least in the baseline revision, there is a bug for collision nodes:

```
scala> case class C(equalityElem: Int)(val other: Int)
defined class C

scala> assert(C(0)(0) == C(0)(1))
```

```
scala> class Key(i: Int) { override val hashCode: Int = 0; override val toString = s"Collide($i)" }
defined class Key

scala> val map = collection.immutable.ChampHashMap[Key, C]().updated(new Key(0), null).updated(new Key(1), null)
map: scala.collection.immutable.ChampHashMap[Key,C] = ChampHashMap(Collide(0) -> null, Collide(1) -> null)

scala> val k2 = new Key(2); map.updated(k2, C(0)(0)).updated(k2, C(0)(1)).apply(k2).other
k2: Key = Collide(2)
res9: Int = 0
```

That comes down to: https://github.com/scala/scala/blob/43bf2351ec0f251f4d6cdc5a0fdac75c7ea6ce1e/src/library/scala/collection/immutable/ChampHashMap.scala#L503-L506


For regular nodes, the overwritten value does appear to win:

```
scala> val map = collection.immutable.ChampHashMap[Key, C]()
map: scala.collection.immutable.ChampHashMap[Key,C] = ChampHashMap()

scala> val k2 = new Key(2); map.updated(k2, C(0)(0)).updated(k2, C(0)(1)).apply(k2).other
k2: Key = Collide(2)
res10: Int = 1
```

With this PR:

```
scala> val map = collection.immutable.ChampHashMap[Key, C]()
map: scala.collection.immutable.ChampHashMap[Key,C] = ChampHashMap()

scala> val k2 = new Key(2); map.updated(k2, C(0)(0)).updated(k2, C(0)(1)).apply(k2).other
k2: Key = Collide(2)
res0: Int = 1

scala> val map = collection.immutable.ChampHashMap[Key, C]().updated(new Key(0), null).updated(new Key(1), null)
map: scala.collection.immutable.ChampHashMap[Key,C] = ChampHashMap(Collide(0) -> null, Collide(1) -> null)

scala> val k2 = new Key(2); map.updated(k2, C(0)(0)).updated(k2, C(0)(1)).apply(k2).other
k2: Key = Collide(2)
res1: Int = 1
```

>  The following snippet would fail with the change above:

That code doesn't typecheck against the `ChampHashMap` API, as only the internal `MapNode` has the `contains(key, value, ..)` API. 

What seems to be missing, before and after this PR, is sharing of the map when the new value really `eq` the exisitng one.

```
scala> case class A(a: Int); val map = collection.immutable.ChampHashMap[Int, A](); val a = A(1); val map1 = map.updated(1, a); val map2 = map1.updated(1, a); map1 eq map2
defined class A
map: scala.collection.immutable.ChampHashMap[Int,A] = ChampHashMap()
a: A = A(1)
map1: scala.collection.immutable.ChampHashMap[Int,A] = ChampHashMap(1 -> A(1))
map2: scala.collection.immutable.ChampHashMap[Int,A] = ChampHashMap(1 -> A(1))
res0: Boolean = false
```